PR for code review

aakahlow opened the pull request.

powerjg committed:
Update gem5sa to work with celery correctly

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update directory to zip results

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Always get hash and dump info.json

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update gem5run result artifact with correct command

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
[in-progress] Add gem5run to the db

aakahlow committed:
Fix results zip file creation

aakahlow committed:
Check database before run

powerjg committed:
Clean up python style in gem5run

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add type to gem5run in the database

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add some functions to search db

These functions will return artifacts or gem5Run instances.

The gem5Run instances aren't tested right now

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Make from gem5sa import * work

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
Fix getRuns and loadFromDict to work with the database

powerjg committed:
Fix json deserialization and other cleanup

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add uuid back to gem5run

oops

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add __str__ to gem5run

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update printing and __repr__ some more

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
remove execute on artifact.py

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
move everything to gem5art subdir

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Added BSD license

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add empty README.md

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add setup.py to package gem5art

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add gitignore

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update the file doc for gem5run.py

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add skeleton for readthedocs

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Restructure gem5art

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Completely change the layout again

This time, we can now publish the gem5art subprojects
separately

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update gitignore for other egg-infos

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update readme to explain the directory structure

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add more info to README

Talk about how to install a bit and discuss how to build
for distribution on PyPI

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add .vscode to gitignore

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add readme's for each package

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Decouple artifact from artifactdb

Add a factory function to get the database connection instead
of implicitly creating it by instantiating an ArtifactDB

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Make artifactdb.py a 'private' module

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add mypy_cache to gitignore

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update artifact to use type annotations

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update run package to use type annotations

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
Fix import error and type annotations

aakahlow committed:
Fixed error with git type

aakahlow committed:
Fix _getSerializable()

aakahlow committed:
Add __hash__ in Artifact() and fix checkArtifacts in run.py

powerjg committed:
Update types in artifact db

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Bump artifact and run versions

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Update gem5Run to use factory pattern

Instead of using inheiritance to have two different constructors for
gem5Run (one for FS and one for SE), use the factory pattern. This
greatly simplifies the code to load from the database fixing a bug where
some members of the class were overwritten by the constructor when
loading from the database.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add mypy config file and fix imports

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add mypy config for run

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add mypy config for tasks

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Fix lots of bugs/type errors in run

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add some simple tests to artifact

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Bump minor version number

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Create CI file for github actions (#3)

powerjg committed:
Add CI badge

powerjg committed:
Create a github action to publish to pypi (#4)

This is just testing for now. I'll test tagging and see if it works.

powerjg committed:
Update publish-to-pypi.yml

powerjg committed:
Update name of pypi action

aakahlow committed:
Update createFSRun

powerjg committed:
Fix setup.py to actually include code

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
Add documentation

powerjg committed:
Bump version to 0.2.1

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
Documentation (#14)

* Update documentation for sphinx and API doc.

* Update formating plus links

aakahlow committed:
Update docs/conf.py (#15)

aakahlow committed:
Read the docs (#16)

* check RTD build

* check RTD build 2

* Add requirements file

* Update requirements

* revert conf.py

* Update links in tutorial

* Update tutorial with github links

powerjg committed:
Update PyPI publish workflow (#17)

I believe we need to be in the local dir for each package to make it so that it will work on PyPI.

powerjg committed:
Update README with info on how to run tests

Closes #10

powerjg committed:
Fix import to have fewer artifacts

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
Read the docs (#18)

* check RTD build

* check RTD build 2

* revert conf.py

* Some updates in the documentation (text + layout)

* Update requirements to render md tables

* Update conf.py to render md tables

* Update celery paths

* Update disk image section

* Update tutorial 1 doc.

* Update the actual boot launch script

* wip on tutorial 2

* minor fix in tutorial1

* updates in tutorial 2 (np)

* update in tutorial2

* Add run script for tutorial 2

* Update config scripts for NPB

* Updates in tutorial 2

* Update configs

* Update npb tutorial

* Update run script for NPB

* Update NPB runscript based on comments

* Update disks.md

* Rename tutorial files

* update in boot tests launch script

* some updates in npb-tutorial based on comments

* Update tutorial titles

* Update npb tutorial based on comments

* more updates in npb tutorial

* update writeBenchScript in run_npb.py

* Add files needed for npb disk creation and installation of npb

* remove .pyc files

* Some updates in boot-test tutorial based on comments

* update gem5 binary artifact's type

* an update in npb-tutorial

* Add npb tests launch script and update boot tests launch script

powerjg committed:
Remove extra import in artifactdb and artifact

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add missing time field to artifact

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
Add copyright to all files (#24)

aakahlow commented on artifact/gem5art/artifact/artifact.py:
```diff
@@ -75,103 +106,139 @@ class Artifact:
             # Now that we have a complete object, construct it
-            self = cls(self)
+            self = cls(data)
             if old_artifact != self:
```
Will this condition ever be true?

powerjg replied:
You're correct! I'll refactor this. See https://gem5.atlassian.net/browse/DARCHR-111

powerjg replied:
Fixed in #26

aakahlow commented on artifact/gem5art/artifact/_artifactdb.py:
```diff
@@ -0,0 +1,90 @@
+        self.artifacts = self.db.artifacts
+        self.fs = gridfs.GridFSBucket(self.db, disable_md5=True)
+
+    def put(self, key: UUID, artifact: Dict[str,Union[str,UUID]]) -> None:
```
Do we fundamentally need to pass a key to the put function, when artifacts.insert_one does not require a key?

powerjg replied:
Fundamentally, no. However, I think it's a clearer API if we do. Note that insert_one *does* require a key, it's just implicit in the dictionary as `_id`. 

I'm open to suggestions, though.

hnpl commented on artifact/gem5art/artifact/artifact.py:
```diff
@@ -75,103 +106,139 @@ class Artifact:
         """Constructs the object from the database based on a UUID or
         dictionary from the database
         """
+        if isinstance(other, str):
```
I guess we need some way to assert (or to document) that "if other is a string then it must be a UUID"

powerjg replied:
If I add "if other is a string it must be constructable as a UUID" to the documentation would that be enough?

hnpl commented on artifact/gem5art/artifact/artifact.py:
```diff
@@ -75,103 +106,139 @@ class Artifact:
         self.git = other['git']
         self.cwd = other['cwd']
-        self.inputs = other['inputs']
+        self.inputs = [Artifact(i) for i in other['inputs']]
```
other['inputs'] is a list of Artifact, but the `other` parameter of `__init__()` does not accept Artifact type

powerjg replied:
other['inputs'] should be a list of UUIDs (or strings which are UUIDs depending on if it's coming from json or bson).

aakahlow commented on run/gem5art/run.py:
```diff
@@ -0,0 +1,516 @@
+                                           run.script_name,
+                                           *run.params)
+
+        run.outdir = os.path.abspath(run.outdir)
```
I think, this should be run.outdir = os.path.abspath(run.relative_outdir)

powerjg replied:
Yep. Fixed in #27!

powerjg committed:
Refactor __eq__ and remove useless check

Refactor __eq__ to only check for equality and move the code that checks
similarity into its own function. This allows us to remove the check for
equality in the factory function and instead call the checkSimilar
function.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Fix createSERun using outdir instead of relative_outdir

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Bump to 0.3.0

Major change is moving around artifact import statements
Many bugs fixed as well

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

aakahlow committed:
More tests (#28)

* add artifact similarity test

* Add test for gem5Run

* Update artifact test to fix import path

* Update CI to install gem5art dependencies

* Add command test in run tests

* update artifact eq & similarity tests

* fix white spaces

powerjg committed:
Add gem5art logo

powerjg committed:
Add logo to README

aakahlow committed:
Minor updates in documentation (DARCHR-109 & DARCHR-110) (#32)

* Update docs to add time attribute of artifact - DARCHR-109

* update requirements to install latest gem5art on RTD server

* Update artifact import paths in docs (DARCHR-110)

* Add docs badge to the README file

* Add gem5art logo

hnpl committed:
Remove channels parameter when calling memory constructor

A recent gem5 commit removes 'channels' parameter from memory
constructor. This commit removes the 'channels' parameter when
calling memory constructor in boot-tests and npb-tests config.

hnpl committed:
Add linux-4.19.83 config

Linux 4.19.83 config is generated from 4.14.134 config as
described in docs/linux-configs/README.md.

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

krithi96 committed:
Krithiga/integration microbenchmark (#38)

* updates to integration_microbenchmark

* launch scirpt

* launch scripts for microbenchmark

* updated working scripts

powerjg committed:
Some updates to the artifact doc page (#36)

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Some minor updates to the intro (#35)

Fixed some style issues and added a few sentences

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
Add some prose to the index page (#34)

This gives some overview of gem5art and why we created it.

powerjg committed:
Try to fix protocol wiki link (#40)

* Try to fix protocol wiki link

* Update  conf and source files to use .md as main syntax

* remove .rst files

* Fix broken links

* Fix the heading level of API doc. sections

* Artifact --> Artifact Module in API doc

aakahlow committed:
Add option to set output path dir (#41)

* Add option to set output path dir

* Update the run test based on the out dir change

aakahlow committed:
bump the version to 0.3.1

aakahlow committed:
Update linux source clone commands (#43)

aakahlow committed:
Update launch scripts with outdir (#42)

* Update launch scripts with outdir

* fix some problems with updates

powerjg committed:
Add the spec2006 tutorial and relevant scripts

Add the spec2006 tutorial and relevant scripts.

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

aakahlow committed:
Update .gitignore files in tutorials (#45)

aakahlow committed:
Update the linux kernel versions in boot and npb tutorials (#44)

MiraWelner committed:
added very beginning of HTML file (#23)

* added begining of website script

* added JSON script to artifacts folder

* Update gem5art getruns to be a standalone script

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

* Remove uneeded function

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg commented:
I'm closing this PR. Whatever code review still needs to happen can happen as we go :)

l00515693 commented on docs/configs-boot-tests/sample.py:
```diff
@@ -0,0 +1,305 @@
+    for i, insts in enumerate(sample_insts):
+        print "Fast forwarding to sample", i, "stopping at", insts
+        insts_past = fastforward(system, insts - system.totalInsts())
+        if insts_past/insts > 0.01:
```
Hi Jason, I wonder if the scheduleInstStop(0, exit_insts, "Max Insts") will result in the exceeded instrutions(real insts > exit_insts) when m5.simulate() stop?

powerjg replied:
I'm not sure if I understand your question. However, I think the comments in the fastforward function are pretty clear: https://github.com/darchr/gem5art/blob/master/docs/gem5-configs/configs-boot-tests/sample.py#L61. When running with multiple CPUs you can only set the `scheduleInstStop` on a per-CPU basis, so it's imprecise if you want to stop after a certain number of global instructions.

In practice, I don't think I've ever seen `insts_past/insts > 0.01`.

l00515693 commented:
you mean that scheduleInstStop function is very accurate for stopping m5.simulate at my expected ROI instruction numbers for one core simulation？


In my practice，for example：when I set scheduleInstStop with 10000000 instrutions，but when the simulation stop，I use totalInsts function to print current instrutions，it showed about 100456789，exceeded 456789 instrutions.



发自我的iPhone


------------------ Original ------------------
From: Jason Lowe-Power <notifications@github.com&gt;
Date: Thu,Aug 6,2020 10:33 PM
To: darchr/gem5art <gem5art@noreply.github.com&gt;
Cc: CharlieYichao <103342943@qq.com&gt;, Comment <comment@noreply.github.com&gt;
Subject: Re: [darchr/gem5art] PR for code review (#20)

powerjg commented:
That's within 1%! :D

BTW, this conversation is more appropriate for the gem5-users mailing list: http://www.gem5.org/mailing_lists/
